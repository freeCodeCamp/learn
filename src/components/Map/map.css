.map-ui {
  height: 100%;
  margin-left: 35px;
}

@media screen, (max-width: 630px) {
  .map-ui {
    margin: 0 auto;
  }
}

.map-ui ul {
  list-style: none;
  color: #006400;
}

.map-title {
  display: flex;
  align-items: center;
  cursor: pointer;
  padding-top: 10px;
  padding-bottom: 20px;
  border-bottom: 1px solid #e2e2e2;
}

.map-title > svg {
  width: 14px;
  margin-right: 5px;
  flex-shrink: 0;
}

li.open > .map-title svg {
  transform: rotate(90deg);
}

.map-challenge-title {
  margin-bottom: 0.25rem;
}

.night .map-challenge-title a {
  color: #f8f8f8;
}

.map-challenge-title-completed {
  opacity: 0.5;
}
