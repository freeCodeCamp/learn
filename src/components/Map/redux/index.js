import { createAction, handleActions } from 'redux-actions';

import { createTypes } from '../../../../utils/stateManagment';

const ns = 'map';

export const getNS = () => ns;

const initialState = {
  expandedState: {
    superBlock: {
      'Responsive Web Design': true
    },
    block: {
      'basic-html-and-html5': true
    }
  }
};

const types = createTypes(['toggleSuperBlock', 'toggleBlock'], ns);

export const toggleBlock = createAction(types.toggleBlock);
export const toggleSuperBlock = createAction(types.toggleSuperBlock);

export const makeExpandedSuperBlockSelector = superBlock => state =>
  !!state[ns].expandedState.superBlock[superBlock];
export const makeExpandedBlockSelector = block => state =>
  !!state[ns].expandedState.block[block];

export const reducer = handleActions(
  {
    [types.toggleBlock]: (state, { payload }) => ({
      ...state,
      expandedState: {
        ...state.expandedState,
        block: {
          ...state.expandedState.block,
          [payload]: !state.expandedState.block[payload]
        }
      }
    }),
    [types.toggleSuperBlock]: (state, { payload }) => ({
      ...state,
      expandedState: {
        ...state.expandedState,
        superBlock: {
          ...state.expandedState.superBlock,
          [payload]: !state.expandedState.superBlock[payload]
        }
      }
    })
  },
  initialState
);
