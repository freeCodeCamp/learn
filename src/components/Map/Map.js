import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import uniq from 'lodash/uniq';

import SuperBlock from './components/SuperBlock';
import Spacer from '../util/Spacer';

import './map.css';
import { ChallengeNode } from '../../redux/propTypes';

const propTypes = {
  introNodes: PropTypes.arrayOf(
    PropTypes.shape({
      fields: PropTypes.shape({ slug: PropTypes.string.isRequired }),
      frontmatter: PropTypes.shape({
        title: PropTypes.string.isRequired,
        block: PropTypes.string.isRequired
      })
    })
  ),
  nodes: PropTypes.arrayOf(ChallengeNode)
};

class ShowMap extends PureComponent {
  renderSuperBlocks(superBlocks) {
    const { nodes, introNodes } = this.props;
    return superBlocks.map(superBlock => (
      <SuperBlock
        introNodes={introNodes}
        key={superBlock}
        nodes={nodes}
        superBlock={superBlock}
      />
    ));
  }

  render() {
    const { nodes } = this.props;
    const superBlocks = uniq(nodes.map(({ superBlock }) => superBlock));
    return (
      <div className='map-ui'>
        <ul>
          {this.renderSuperBlocks(superBlocks)}
          <Spacer />
        </ul>
      </div>
    );
  }
}

ShowMap.displayName = 'Map';
ShowMap.propTypes = propTypes;

export default ShowMap;
