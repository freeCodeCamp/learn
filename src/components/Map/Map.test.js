/* global expect */

import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Map from './Map';
import mockNodes from '../../__mocks__/map-nodes';
import mockIntroNodes from '../../__mocks__/intro-nodes';

Enzyme.configure({ adapter: new Adapter() });
const renderer = new ShallowRenderer();

test('<Map /> snapshot', () => {
  const component = renderer.render(
    <Map introNodes={mockIntroNodes} nodes={mockNodes} />
  );
  expect(component).toMatchSnapshot('Map');
});
