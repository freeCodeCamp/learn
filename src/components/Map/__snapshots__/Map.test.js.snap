// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> snapshot: Map 1`] = `
<div
  className="map-ui"
>
  <ul>
    <Connect(SuperBlock)
      introNodes={
        Array [
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-a",
            },
            "frontmatter": Object {
              "block": "Block A",
              "title": "Introduction to Block A",
            },
          },
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-b",
            },
            "frontmatter": Object {
              "block": "Block B",
              "title": "Introduction to Block B",
            },
          },
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-c",
            },
            "frontmatter": Object {
              "block": "Block C",
              "title": "Introduction to Block C",
            },
          },
        ]
      }
      nodes={
        Array [
          Object {
            "block": "block-a",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-one",
            },
            "id": "a",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-two",
            },
            "id": "b",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-one",
            },
            "id": "c",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-two",
            },
            "id": "d",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-c",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block C",
              "slug": "/super-block-one/block-c/challenge-one",
            },
            "id": "e",
            "isPrivate": true,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-one",
            },
            "id": "f",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-two",
            },
            "id": "g",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-one",
            },
            "id": "h",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge One",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-two",
            },
            "id": "i",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge Two",
          },
        ]
      }
      superBlock="Super Block One"
    />
    <Connect(SuperBlock)
      introNodes={
        Array [
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-a",
            },
            "frontmatter": Object {
              "block": "Block A",
              "title": "Introduction to Block A",
            },
          },
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-b",
            },
            "frontmatter": Object {
              "block": "Block B",
              "title": "Introduction to Block B",
            },
          },
          Object {
            "fields": Object {
              "slug": "/super-block-one/block-c",
            },
            "frontmatter": Object {
              "block": "Block C",
              "title": "Introduction to Block C",
            },
          },
        ]
      }
      nodes={
        Array [
          Object {
            "block": "block-a",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-one",
            },
            "id": "a",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-two",
            },
            "id": "b",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-one",
            },
            "id": "c",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-two",
            },
            "id": "d",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-c",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block C",
              "slug": "/super-block-one/block-c/challenge-one",
            },
            "id": "e",
            "isPrivate": true,
            "isRequired": false,
            "superBlock": "Super Block One",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-one",
            },
            "id": "f",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge One",
          },
          Object {
            "block": "block-a",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block A",
              "slug": "/super-block-one/block-a/challenge-two",
            },
            "id": "g",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge Two",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-one",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-one",
            },
            "id": "h",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge One",
          },
          Object {
            "block": "block-b",
            "dashedName": "challenge-two",
            "fields": Object {
              "blockName": "Block B",
              "slug": "/super-block-one/block-b/challenge-two",
            },
            "id": "i",
            "isPrivate": false,
            "isRequired": false,
            "superBlock": "Super Block Two",
            "title": "Challenge Two",
          },
        ]
      }
      superBlock="Super Block Two"
    />
    <Spacer />
  </ul>
</div>
`;
